### ISSUE TITLE
osqueryd: metrics

### ISSUE THREAD

**Issue by SuperQ** [facebook/osquery#5541]
# Feature request

We have osqueryd deployed on our fleet, but it doesn't seem to have any good monitoring support. It would be useful to have a Prometheus/OpenMetrics endpoint in the daemon so we can monitor the state of the service.

### What new feature do you want?

Add a Prometheus/OpenMetrics HTTP endpoint to osqueryd.

### How is this new feature useful?

Easy monitoring of osqueryd internals and general process metrics.

### How can this be implemented?

There is a C++ library for Prometheus metrics: https://github.com/jupp0r/prometheus-cpp

**Comment by directionless** [comment]
This has come up a couple of times. Here, in office hours ([notes](https://hackmd.io/OpICAxxJTkGNAH7rzX5Wjw), [video](https://youtu.be/GbXi6xR7gZE)). I wanted to put a couple of thoughts down.

I think this is an interesting thing to explore, but I see some bit question to resolve first. As I understand it. prometheus operates on a pull model. Endpoints track metrics, generally simple numbers. And prometheus scrapes them. In contrast, osquery operates on a push model, and what it pushes is pretty arbitrary json.

This highlights two mismatches.

First, the push vs pull one. Osquery generally follows a push model, vs prometheuses pull model. We are generally going to be very hesitant to move osquery to a listening daemon. That creates a host of security problems.

Second, the data typing. I think of prometheus as something like a straight metrics system. Not a log processor. But osquery generates something closer to json logs. While I can imagine writing queries that produced output that more closely matches prometheus, I think it’s a fairly small niche.

Both of these issues may be better solved by using an intermediary to handle the log ingest, and then export to prometheus. Not totally sure though, if there’s a common export endpoint, it might be pretty small work to create a log plugin that can push data to it.

**Comment by SuperQ** [comment]
In order to have good monitoring, you need to check the health state of each process.

For example, I know I have osqueryd installed on a host. How do I know that it's able to send push messages to the remote collector? How do I know it's operating correctly?

In the Prometheus monitoring world, we scrape a simple http endpoint to collect simple metrics about the internal state of a process. "How many messages have I pushed?" "How many message errors?" "How many message retries?" Whatever you need in order to keep track of things at a per-process level.

Using an intermediary doesn't allow us to know that something between the target daemon and the intermediary is broken. What if a process is configured to send to the wrong one? What if there's a protocol mismatch? What if there's accidentally a new firewall rule in place?

Prometheus endpoints solve this because it's an active polling monitoring system. I know because my configuration source of truth (service discovery) tells me that there must be an osqueryd installed. If Prometheus can't reach that process, I can fire an alert. If Prometheus alerting rules evaluate that "The process is having a lot of errors", an alert can be fired.

This isn't about sending the entirety of the event logging data that osqueryd is sending to the collector. In fact, that's _not_ what we want. It's just about being able to capture the internal state of the running process via external means. The only trick is the "external means" is a simple HTTP API endpoint.

**Comment by directionless** [comment]
> This isn't about sending the entirety of the event logging data that osqueryd is sending to the collector. In fact, that's _not_ what we want. It's just about being able to capture the internal state of the running process via external means. The only trick is the "external means" is a simple HTTP API endpoint.

Ah. I think you're saying you're interested in using prometheus to monitor _osquery_. Not that you're interested in using osquery to push data into prometheus? (The latter is the usual way this comes up)

This seems like a reasonable use case, but still bumps into that first impedance mismatch. Though if you reduce the data served to a handful of internal metrics makes it seems more reasonable.

**Comment by SuperQ** [comment]
Yes, that's exactly it. Prometheus is a monitoring tool meant to replace things like Nagios.

Sadly, I don't actually use osquery anymore. But maybe someone else.

**Comment by stefanmaerz** [comment]
+1 on this. Having prom export of osquery health metrics would be fantastic for an operational perspective.
